File: src/main/java/cz/vhromada/catalog/gui/genre/GenresStatsTableDataModel.java
package cz.vhromada.catalog.gui.genre;

import java.util.List;

import cz.vhromada.catalog.entity.Genre;
import cz.vhromada.catalog.facade.GenreFacade;
import cz.vhromada.catalog.gui.commons.AbstractStatsTableDataModel;
import cz.vhromada.result.Result;
import cz.vhromada.result.Status;

import org.springframework.util.Assert;

/**
 * A class represents data model for table with stats for genres.
 *
 * @author Vladimir Hromada
 */
public class GenresStatsTableDataModel extends AbstractStatsTableDataModel {

    /**
     * SerialVersionUID
     */
    private static final long serialVersionUID = 1L;

    /**
     * Facade for genres
     */
    private final GenreFacade genreFacade;

    /**
     * List of genres
     */
    private List<Genre> genres;

    /**
     * Creates a new instance of GenresStatsTableDataModel.
     *
     * @param genreFacade facade for genres
     * @throws IllegalArgumentException if facade for genres is null
     */
    public GenresStatsTableDataModel(final GenreFacade genreFacade) {
        Assert.notNull(genreFacade, "Facade for genres mustn't be null.");

        this.genreFacade = genreFacade;
        update();
    }

    @Override
    public Object getValueAt(final int rowIndex, final int columnIndex) {
        switch (columnIndex) {
            case 0:
                return genres.size();
            default:
                throw new IndexOutOfBoundsException("Bad column");
        }
    }

    @Override
    public int getColumnCount() {
        return 1;
    }

    @Override
    public int getRowCount() {
        return 1;
    }

    @Override
    public Class<?> getColumnClass(final int columnIndex) {
        return Integer.class;
    }

    @Override
    public String getColumnName(final int column) {
        switch (column) {
            case 0:
                return "Count of genres";
            default:
                throw new IndexOutOfBoundsException("Bad column");
        }
    }

    @Override
    public final void update() {
        final Result<List<Genre>> result = genreFacade.getAll();

        if (Status.OK == result.getStatus()) {
            genres = result.getData();
        } else {
            throw new IllegalArgumentException("Can't get data. " + result);
        }
    }

}

File: src/main/kotlin/cz/vhromada/catalog/gui/genre/GenresStatsTableDataModel.kt
package cz.vhromada.catalog.gui.genre

import cz.vhromada.catalog.entity.Genre
import cz.vhromada.catalog.facade.GenreFacade
import cz.vhromada.catalog.gui.common.AbstractStatsTableDataModel
import cz.vhromada.validation.result.Status

/**
 * A class represents data model for table with stats for genres.
 *
 * @author Vladimir Hromada
 */
class GenresStatsTableDataModel(private val genreFacade: GenreFacade) : AbstractStatsTableDataModel() {

    /**
     * List of genres
     */
    private lateinit var genres: List<Genre>

    init {
        update()
    }

    override fun getValueAt(rowIndex: Int, columnIndex: Int): Any {
        when (columnIndex) {
            0 -> return genres.size
            else -> throw IndexOutOfBoundsException("Bad column")
        }
    }

    override fun getColumnCount(): Int {
        return 1
    }

    override fun getRowCount(): Int {
        return 1
    }

    override fun getColumnClass(columnIndex: Int): Class<*> {
        return Int::class.java
    }

    override fun getColumnName(column: Int): String {
        when (column) {
            0 -> return "Count of genres"
            else -> throw IndexOutOfBoundsException("Bad column")
        }
    }

    override fun update() {
        val result = genreFacade.getAll()

        if (Status.OK == result.status) {
            genres = result.data!!
        } else {
            throw IllegalArgumentException("Can't get data. $result")
        }
    }

}
